### DISCUSSION TITLE
Idea: Primary Unnamed Constructor Initializer for C#

### DISCUSSION THREAD

**Discussion by Mubarrat** [dotnet/csharplang#8219] [Language Ideas]
Hello, C# language enthusiasts,

I would like to share an idea for a new feature in the C# language: the Primary Unnamed Constructor Initializer. This feature aims to streamline object initialization by allowing methods or events to be set directly in the constructor initializer, providing a more concise and readable syntax. You can find the concept code and more details in my repository here: [Primary Unnamed Constructor Initializer Concept](https://github.com/Mubarrat/CSharp-Concepts/blob/main/PrimaryUnnamedConstructorInitializer/concept.cs).

### Motivation:

The Primary Unnamed Constructor Initializer builds upon the existing primary constructor concept by enabling more flexible and expressive object initialization. This feature would improve code readability and maintainability, especially when dealing with complex initialization logic. This would be the successor of the Primary Constructor introduced in C# 12.

### Idea:

The primary unnamed constructor initializer can be placed flexiblely. Here is an example to illustrate its usage:

```csharp
// Concept class
public class Sample(bool startImmediately = false) : public() // "public()" here isn't necessary but it can be used to define constructor access. I think it should come first before any base class.
{
    // Unnamed Initializer. Can be used to execute flexiblely.
    OnStart += () => Logger.Log("Application Started");
	
    public event Action OnStart;
    public ILogger Logger { get; } = new Logger();

    // It will be executed after Logger property defined.
    if (startImmediately)
	Start();

    public void Start() => OnStart?.Invoke();
}

public interface ILogger
{
    void Log(string message);
}

public class Logger : ILogger
{
    public void Log(string message)
        => Console.WriteLine($"Log: {message}");
}

public class Program
{
    public static void Main()
    {
        var sample = new Sample();
        sample.Start();
    }
}
```

### Explanation:

- The first independent statement executed first.
- Defined Logger property.
- Second Statements executed after property defined.
- The `Start` method invokes the `OnStart` event, demonstrating how the initializer sets up the necessary functionality.

### Benefits:

- **Enhanced Readability**: By consolidating initialization logic into the constructor initializer, the code becomes more readable and easier to follow.
- **Improved Maintainability**: This approach reduces the likelihood of errors by ensuring that essential methods and events are set up at the beginning of the object's lifecycle.
- **Concise Syntax**: The primary unnamed constructor initializer allows for a more concise and expressive syntax, reducing boilerplate code.

### Contribution:

I encourage the community to review the concept and provide feedback on its potential inclusion in the C# language. Let's discuss how this feature could enhance our coding experience and bring additional clarity to object initialization in C#.

Thank you for considering this idea.

**Comment by HaloFour** [comment]
How would `public()` work given that's where you'd already put a `base(...)` constructor call?

As for the initializer, I would rather that it appear in some kind of block and not simply as part of the class body as the latter could easily be ambiguous.  This was a big point of contention with primary constructors as they were first discussed back in the C# 6.0 timeframe (unfortunately those conversations are buried in the CodePlex archive so it's not possible to easily link to them.)

**Reply by Mubarrat** [reply]
Consider this:

```csharp
public class DerivedClass(...) : public(), BaseClass(/* You can pass parameter here */)
{
    ...
}
```

**Reply by Mubarrat** [reply]
> As for the initializer, I would rather that it appear in some kind of block and not simply as part of the class body as the latter could easily be ambiguous.

This kind of code, unnamed initializer must come before any field or property is defined.

**Reply by HaloFour** [reply]
> must come before any field or property is defined.

That would still be ambiguous as initialized local variables and local functions are effectively indistinguishable from fields or methods.

Even if the language had rules to disambiguate, I still think it would be worthwhile to the reader to explicitly separate the initializer into a block.  C# could borrow a page from Java initializer blocks or record constructors here, or introduce a new form of syntax such as the `init` block:

```cs
public class Foo(int x) {
    init {
        if (x < 0) throw new ArgumentOutOfRangeException(nameof(x));
    }
}
```

Syntax aside, I do believe that the language team does find value in the concept.  The questions are around whether such a block should execute during construction, after initialization (after all `init` members are set), or if both forms should be offered.

**Reply by HaloFour** [reply]
See:
Final Initializer discussion #6591

**Reply by Mubarrat** [reply]
Might be this:

```csharp
// Concept class
public class Sample(bool startImmediately = false) : public()
{
    // Unnamed Initializer. Can be used to execute flexiblely.
    OnStart += () => Logger.Log("Application Started");
	
    public event Action OnStart;
    public ILogger Logger { get; } = new Logger();

    // It will be executed after Logger property defined.
    if (startImmediately)
	Start();

    public void Start() => OnStart?.Invoke();
}

public interface ILogger
{
    void Log(string message);
}
```

Translation Code:

```csharp
// Concept class
public class Sample()
{
    public Sample(bool startImmediately = false)
    {
        OnStart += () => Logger.Log("Application Started");
        Logger = new Logger();
        if (startImmediately)
	    Start();
    }

    public event Action OnStart;
    public ILogger Logger { get; }

    public void Start() => OnStart?.Invoke();
}

public interface ILogger
{
    void Log(string message);
}
```

**Reply by Mubarrat** [reply]
@HaloFour, I need to define flexiblely. If we use fixed place, it would break the concept.

**Reply by HaloFour** [reply]
>  I need to define flexiblely. If we use fixed place, it would break the concept.

I don't understand this statement.  You just stated that the unnamed initializer must come before member declarations for the sake of avoiding ambiguity.  How do you achieve both goals?

Field initializers already occur prior to the constructor body, which means that they cannot access any instance members.  That already means that this flexibility you want is not possible, and changing field initialization would be a massive breaking change to the language.

**Reply by HaloFour** [reply]
Initializer blocks could be mixed between field initializers and executed lexically, but only if the initializer blocks were limited in that they cannot invoke any instance members aside from fields.  That would satisfy the rule that they execute prior to the base constructor class while retaining lexical order.  However, that also means that it would be illegal for an initialization block to call the `Start` method, as in your examples.

Also, that doesn't prevent relegating those statements to blocks as the proposal could include allowing more than one block to be declared.  This is how initialization blocks already work in Java.

**Reply by Mubarrat** [reply]
Sorry, I changed my mind. I want flexibility. See the other comment. Unnamed initializer now doesn't need to come first.

**Reply by Mubarrat** [reply]
The problem is how we can understand difference between field and variable. Since, initializers are flexible to class block, not in constructor block. So, there is no variable except in code blocks. Now what is the rule breaking changes?

**Reply by Mubarrat** [reply]
Now, the following concept example is:

```csharp
// Concept class
public class Sample(bool startImmediately = false) : public()
{
    // Unnamed Initializer. Can be used to execute flexiblely.
    OnStart += () => Logger.Log("Application Started");
	
    public event Action OnStart;
    public ILogger Logger { get; } = new Logger();

    // It will be executed after Logger property defined.
    if (startImmediately)
	Start();

    public void Start() => OnStart?.Invoke();
}
```

Why using `Start()` is illegal. Start() is a void method? This would be illegal if we define it using return type `Action` field. You know that method can be defined anywhere in a class and can be called anywhere from same-class.

**Reply by HaloFour** [reply]
Initializers execute prior to the base constructor call, and as such they are forbidden from calling instance members.  At most they can set instance fields, but they cannot call methods or property accessors.  Given field initializers always execute prior to the base constructor call, in order to support lexical evaluation of "initializer" bodies intermingled with field initializers, they would have to execute at the same time.  Thus, calling instance methods would be forbidden.

This restriction is because an arbitrary method may be able to invoke virtual members declared on the type, which could be overridden in a derived type, and if those overridden methods are called they could observe the state of the class prior to construction.  This is why the runtime and language both forbid referencing instance members, aside from declared fields, in the constructor prior to the base constructor call.

**Reply by Mubarrat** [reply]
Sorry, I'm here not to change syntax. I wanted to create a new syntax which relays on existing syntax. I'll see replies tomorrow.

**Comment by FaustVX** [comment]
There is already a Final Initializer proposal #6591 
I think it's better to have proper block of code rather than having code spread all over the class body.

**Reply by Mubarrat** [reply]
This kind of code, unnamed initializer or whatever you say must come before any field or property is defined.

**Reply by FaustVX** [reply]
I really don't think than having code directly at the top of the class body is readable.

**Reply by Mubarrat** [reply]
https://github.com/dotnet/csharplang/blob/5055b97eee8c10d12f822f6d4db9464329615947/proposals/final-initializers.md?plain=1#L20
According to your link reference, it says it'll execute after all constructors or initializers. But It would be ideal if it were executed before any property or field. Sometimes, we've to throw `ArgumentNullException` if the parameters are null. What do you do? Do you throw exceptions after defining a property or field, or before?

**Reply by Mubarrat** [reply]
Might be this:

```csharp
// Concept class
public class Sample(bool startImmediately = false) : public()
{
    // Unnamed Initializer. Can be used to execute flexiblely.
    OnStart += () => Logger.Log("Application Started");
	
    public event Action OnStart;
    public ILogger Logger { get; } = new Logger();

    // It will be executed after Logger property defined.
    if (startImmediately)
	Start();

    public void Start() => OnStart?.Invoke();
}

public interface ILogger
{
    void Log(string message);
}
```

Translation Code:

```csharp
// Concept class
public class Sample()
{
    public Sample(bool startImmediately = false)
    {
        OnStart += () => Logger.Log("Application Started");
        Logger = new Logger();
        if (startImmediately)
	    Start();
    }

    public event Action OnStart;
    public ILogger Logger { get; }

    public void Start() => OnStart?.Invoke();
}

public interface ILogger
{
    void Log(string message);
}
```

**Reply by FaustVX** [reply]
No, you'll change current behavior.
Current is, when you create an object :
1. field/Property initializer
2. base object creation (recursive)
3. constructor

So, if you move the step 1 inside the ctor, you'll introduce breaking changes

[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsAGAAizwDoAldBKMAU2oGEB7MABygBtmyAMqCAblADGzAM4BuYsRjMA7uTYAKAJTkA3mrzkAvOQCC1APKoE3K1vIBfeUWK5TxHcXJfKAZnKwEUwNjS2tbTSdvck9vLD8TLRior0MAPnJQmwR1ACIzCQQOZByIpOSyuMo8ADZ/GEDM2wBtNgg+AWQAWWYwYEEAOQgWAF0qCmkEZm4jchhUdoBCTTKPImTYvABOdQmp0rX1ygB2clJI73tiS+cSHHIAIXIQNyJVqMqAh+CMqyytc68FT890SB3WaR+YWyOXu1AKRRKTmuLjubCeD3cQLqgTY30a2X2USxGmWYOSEPxuTYcMKxX29iAA)

That's why Final Initialzer come after the ctor

**Reply by Mubarrat** [reply]
Yes, then there will be no task in step 1.
This will become more flexible.
This initializer is part of the constructor, not like the final initializer or after the task of constructor.
The proposal final initializer is a different concept than I want to explain.
Didn't you understand my concept?
We normally do task in constructor if we don't use Field/Property initializer

**Reply by HaloFour** [reply]
See: #1026
